lib/spack/spack/util/timer.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""Debug signal handler: prints a stack trace and enters interpreter.
+
+``register_interrupt_handler()`` enables a ctrl-C handler that prints
+a stack trace and drops the user into an interpreter.
+
+"""
+import spack.util.spack_json as sjson
+import time
+import sys
+
+
+class Timer(object):
+    """
+    Simple timer for timing phases of a solve or install
+    """
+    def __init__(self):
+        self.start = time.time()
+        self.last = self.start
+        self.phases = {}
+        self.end = None
+
+    def phase(self, name):
+        last = self.last
+        now = time.time()
+        self.phases[name] = now - last
+        self.last = now
+
+    @property
+    def _spack_timer_total(self):
```

Comment:
Ok I see what happened here -- I mean something like this:

```python
_total_name = "_spack_timer_total"

class Timer(object):
    ...
    @property
    def total(self):
        return self.phases[_total_name]
```

The point is to avoid duplicating the logic for computing phase time, and also to allow anyone using this to have a phase called "total".

Reply:
Oh so you want a phase called a derivative of total. Why can’t we keep it separate? It’s not a phase.

Reply:
The issue I see with making the derivative of total a phase is if there is ever some functionality added to operate over phases we will have to do custom logic to skip the total. Anyone parsing the data would have to do the same. I think it makes sense to keep it separate, or even just to have the total derivation be a function over summing up the total of the phases.